### PR TITLE
[DataGrid] Fix calling onCellEditStop on error

### PR DESCRIPTION
**Pull request by sai6855**
closes: https://github.com/mui/mui-x/issues/11406

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

File: packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts [hunk 1]
```diff
@@ -234,6 +234,42 @@ export const useGridCellEditing = (
...
+        let newProps: GridEditCellProps = {
+          ...editingState[id][field],
+          value: parsedValue,
+          changeReason: debounceMs ? 'debouncedSetEditCellValue' : 'setEditCellValue',
+        };
+
+        if (column.preProcessEditCellProps) {
+          const hasChanged = value !== editingState[id][field].value;
+
+          newProps = { ...newProps, isProcessingProps: true };
```
Comment (sai6855) [inline packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts:260]: This entire logic was copied from https://github.com/mui/mui-x/blob/aed8aeb903ef7d319419a7c50eb44286d547888d/packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts#L449-L469

Thought of storing `{ id, row, props: newProps, hasChanged }` in ref, which are computed in below function, but wasn't sure if `setCellEditingEditCellValue` is run always and runs before `runIfNoPreProcessError` https://github.com/mui/mui-x/blob/aed8aeb903ef7d319419a7c50eb44286d547888d/packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts#L445

Comment (cherniavskii) [reply packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts:260]: Thanks!
Did you consider following the same approach as in https://github.com/mui/mui-x/pull/11383

Comment (sai6855) [reply packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts]: Wasn't aware of that PR. Now I've updated code to use same approach as in https://github.com/mui/mui-x/pull/11383.

File: packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts [hunk 2]
```diff
@@ -234,6 +234,19 @@ export const useGridCellEditing = (
     [apiRef],
   );
 
+  const runIfNoFieldErrors =
+    <Args extends any[]>(callback?: (...args: Args) => void) =>
```
Comment (cherniavskii) [inline packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts:238]: Nitpick: better type safety
```suggestion
    <Args extends Parameters<GridEventListener<'cellEditStop'>>>(callback?: (...args: Args) => void) =>
```

**Review by cherniavskii — APPROVED** [review]
Looks good, thanks!

**Pull request merged**